datafusion/expr/src/type_coercion/binary.rs

```diff
@@ -623,6 +640,17 @@ fn null_coercion(lhs_type: &DataType, rhs_type: &DataType) -> Option<DataType> {
     }
 }
 
+fn check_time_unit(datatype: DataType) -> Option<TimeUnit> {
+    use arrow::datatypes::DataType::*;
+    match datatype {
+        Time32(TimeUnit::Second) => Some(TimeUnit::Second),
+        Time32(TimeUnit::Millisecond) => Some(TimeUnit::Millisecond),
+        Time64(TimeUnit::Microsecond) => Some(TimeUnit::Microsecond),
+        Time64(TimeUnit::Nanosecond) => Some(TimeUnit::Nanosecond),
+        _ => None,
+    }
+}
```

Comment:
nitpick: I think this can be written a bit more idiomatically:

```suggestion
fn is_time_with_valid_unit(datatype: DataType) -> bool {
    use arrow::datatypes::DataType::*;
    match datatype {
        Time32(TimeUnit::Second) | Time32(TimeUnit::Millisecond) | Time64(TimeUnit::Microsecond) | TimeUnit::Nanosecond) => true
        _ => false,
    }
}
```